### ISSUE TITLE
RA does not work at all when offline

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-analyzer#12499]
I tried to do a bit of coding on rustc while on a plane without internet, and noticed that RA handles this situation quite badly. `cargo metadata` fails because it cannot refresh the index, and then basically nothing works, even though everything needed should be available offline. In particular, not even "check on save" does anything -- I had to run `./x.py check` by hand. That worked perfectly fine, demonstrating that RA could also have run `x.py` as usual and show the diagnostics in the editor.

**rust-analyzer version**: ad6810e90 2022-06-06 stable

**rustc version**: rustc 1.63.0-nightly (e09449220 2022-05-31)

**relevant settings**:
```json
{
    "rust-analyzer.checkOnSave.enable": true,
    "rust-analyzer.checkOnSave.overrideCommand": [
        "./x.py",
        "check",
        "--json-output",
        "library/std",
        "compiler/rustc",
    ],
    "rust-analyzer.rustfmt.overrideCommand": [
        "./build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt",
        "--edition=2021"
    ],
    "rust-analyzer.procMacro.enable": false,
    "editor.formatOnSave": false,
    "files.watcherExclude": {
        "*rustc*/src/llvm-project/**": true,
        "*rustc*/build/**": true,
    },
    "files.exclude": {
        "src/llvm-project/**": true,
        "build/**": true
    },
    "rust-analyzer.cargo.buildScripts.enable": false,
    "rust-analyzer.rustc.source": "./Cargo.toml",
    "rust-analyzer.hover.actions.references.enable": true,
}
```

**Comment by jhgg** [comment]
Does it work if you add `net.offline = true` to your [cargo config](https://doc.rust-lang.org/cargo/reference/config.html)?

**Comment by RalfJung** [comment]
That only changes the error message to
```
rust-analyzer failed to load workspace: Failed to read Cargo metadata from Cargo.toml file /home/r/src/rust/rustc/Cargo.toml, cargo 1.63.0-nightly (39ad1039d 2022-05-25): Failed to run `"cargo" "metadata" "--format-version" "1" "--manifest-path" "/home/r/src/rust/rustc/Cargo.toml" "--filter-platform" "x86_64-unknown-linux-gnu"`: `cargo metadata` exited with an error: error: failed to download `minifier v0.2.1`

Caused by:
  attempting to make an HTTP request, but --offline was specified
```

**Comment by jhgg** [comment]
Hmm, I would think that if the local cache has all the crates in them it should work fine...

**Comment by RalfJung** [comment]
This might be related to https://github.com/rust-lang/cargo/issues/10096, where `cargo metadata` insists to do some work that is IMO entirely unnecessary.

I see two orthogonal issues here:
- `cargo metadata` failed when it possibly should not have
- RA completely bails when `cargo metadata` fails, even though I would think it can still run `./x.py check` and render the resulting diagnostics in the editor

**Comment by flodiebold** [comment]
> RA completely bails when cargo metadata fails, even though I would think it can still run ./x.py check and render the resulting diagnostics in the editor

Without Cargo metadata, RA doesn't have a project and basically can't do anything. Sure, in theory it could blindly run check to report errors, but 1) check-on-save is really a bolted-on-top workaround to the fact that RA doesn't have most diagnostics built-in yet, so we don't really want to make fundamental changes to the architecture based on it, and 2) I think in most cases, if `cargo metadata` fails, you'd expect `cargo check` to fail as well. This seems like a rather special case.

**Comment by lnicola** [comment]
Should there be an `x.py metadata`? :smile:

**Comment by blaa** [comment]
That's not only "I'm on an airplane" case, but also "I work in airgapped environment and would like to use Rust". We do build docker image with all cargo dependencies on a networked host and then move it inside environment without network access. 

I believe analyzer worked ok about 9 months ago, but we could have used older RA back then too.

**Comment by kting28** [comment]
I think the issue is that when `CargoWorkspace::fetch_metadata()` is called, `--offline` is not passed. I briefly verified in an offline environment. Not clear to me how RA would know it's supposed to be offline. May be it needs to read the `offline = true` flag under `[net]` in `Cargo.toml` first...

**Comment by lnicola** [comment]
Note that even `--offline` won't help in most cases.

**Comment by kting28** [comment]
> Note that even `--offline` won't help in most cases.

right, it only works if things are in the cargo cache before the internet access is removed.  Have you tried setting the env var  `CARGO_NET_OFFLINE=true` ? `cargo metadata` seems to honor this.

**Comment by djc** [comment]
FWIW, I ran into this yesterday on an airplane as well. I managed to work around this in a few ways. I managed to work around this in the following way:

* Quit the editor (in my case, VS Code)
* Kill any "stuck" RA processes
* Run `cargo check --all-targets --offline` in a terminal to update the `Cargo.lock` file
* Then restart the editor

(Separately, editing dependencies required some manual dexterity, but that's more of a Cargo issue.)

In my mind the ideal solution here would be that RA uses some connectivity canary (like a trivial DNS request to Google or Cloudflare DNS servers) to assess whether it should try applying `--offline`/`CARGO_NET_OFFLINE`.

**Comment by lnicola** [comment]
> In my mind the ideal solution here would be that RA uses some connectivity canary (like a trivial DNS request to Google or Cloudflare DNS servers) to assess whether it should try applying --offline/CARGO_NET_OFFLINE.

If we do (#15146), it should be disabled by default, because everyone will go up in arms about it ("rust-analyzer is sending your personal information to Cloudflare!").

**Comment by djc** [comment]
Well, I'm fine with a HTTP request to `rust-lang.org` if we prefer that, instead.

**Comment by OSA413** [comment]
It also fails to start analyzing the files if it can't get dependencies, though the project is compiling just fine.

![image](https://github.com/rust-lang/rust-analyzer/assets/16105352/ccf07f62-7ed4-46c7-a8bd-0037c38c3723)

**Comment by Veykril** [comment]
Realized there is a `--no-deps` flag which prevents metadata from accessing the registry entirely, so with https://github.com/rust-lang/rust-analyzer/pull/17915 the workspace itself will at least load, while no crates-io dependencies will resolve. Not ideal, but better than not working at all

**Comment by Krysme** [comment]
> Realized there is a `--no-deps` flag which prevents metadata from accessing the registry entirely, so with [#17915](https://github.com/rust-lang/rust-analyzer/pull/17915) the workspace itself will at least load, while no crates-io dependencies will resolve. Not ideal, but better than not working at all

Well, this commit breaks my project that builds offline with `cargo vendor`. I believe that it detects if my machine were online, and it finds out that it is not, then it adds the `--no-deps` flag. So before the commit, my `rust-analyzer` works fully, and now it works partially that it cannot find some of the standard library facilities like `String`.

**Comment by Veykril** [comment]
No it doesn't, it retries with `--no-deps` if the first try failed it, so something else is breaking for you.

**Comment by jbethune** [comment]
> May be it needs to read the `offline = true` flag under `[net]` in `Cargo.toml` first...

Reading the flag can be useful for projects that are *always* developed offline. But it is not a useful for projects where some of the maintainers sometimes are offline, because they are sometimes traveling.

It is probably better to check the error status of `cargo metadata` and if it appears that we are not connected to the internet, then RA should try to work with the previously downloaded data.

**Comment by Veykril** [comment]
We can only work with data that `cargo metadata` hands us which is the main problem with offlien scenarios. If you are offline, cargo metadata likely fails without producing any usable data.

**Comment by Krysme** [comment]
OK, I'll try to figure it out, but I've bisected the commits and I'm confident that this commit is what breaks my environment, I'll try providing some hints after I gather more information. And thanks for your reply.

**Comment by zicklag** [comment]
With Rust analyzer version 0.3.2509 for VSCode I was unable to use my typical workflow for working offline.

Before #20018 was merged I was able to connect to the internet, run `cargo metadata`, and as long as my crate versions didn't change I was able to continue using it offline.

I'm on a restricted machine where the main GUI desktop can't connect to the internet, but I can open up a shell that has internet access and I tend to run `cargo fetch` in that shell to get everything downloaded before using the GUI offline.

The issue now is that rust-analzyer is creating a new folder at `target/rust-analyzer/metadata` and since I haven't run `cargo metadata` in _that_ folder ( since it doesn't exist until running the GUI with an internet connection ) the metadata isn't cached and rust-analyzer can't startup.

I realize that this is a little bit of a special case, but it still feels like maybe I should be able to have an option to run rust-analyzer against the normal `cargo metadata` location.

**Comment by Storyyeller** [comment]
Having an option to configure RA to pass `--offline` is not ideal (due to cases of people switching between online and offline as observed earlier in this thread), but it would be better than nothing. Right now, it seems like there's no way to make RA work offline at all. 

In my case, I'm using `cargo vendor`, so building works just fine, but there's no way to get RA to add `--offline` to the cargo metadata calls. Since it infinitely spins unsuccessfully trying to connect to crates.io, this also blocks the package lockfile and prevents running other cargo commands either, forcing me to disable RA entirely.

**Comment by ShoyuVanilla** [comment]
> Having an option to configure RA to pass `--offline` is not ideal (due to cases of people switching between online and offline as observed earlier in this thread), but it would be better than nothing. Right now, it seems like there's no way to make RA work offline at all.
> 
> In my case, I'm using `cargo vendor`, so building works just fine, but there's no way to get RA to add `--offline` to the cargo metadata calls. Since it infinitely spins unsuccessfully trying to connect to crates.io, this also blocks the package lockfile and prevents running other cargo commands either, forcing me to disable RA entirely.

I agree with you that we need some improvements here but you can add [`CARGO_NET_OFFLINE=true`](https://doc.rust-lang.org/cargo/reference/config.html#netoffline) into [`rust-analyzer.cargo.extraEnv`](https://rust-analyzer.github.io/book/configuration.html#cargo.extraEnv) as a workaround for now

**Comment by Storyyeller** [comment]
> I agree with you that we need some improvements here but you can add [`CARGO_NET_OFFLINE=true`](https://doc.rust-lang.org/cargo/reference/config.html#netoffline) into [`rust-analyzer.cargo.extraEnv`](https://rust-analyzer.github.io/book/configuration.html#cargo.extraEnv) as a workaround for now

I already tried that, but it had no effect. Maybe I did something wrong?

**Comment by ShoyuVanilla** [comment]
The config would look like the following in vs-code

```jsonc
// settings.json
{
    "rust-analyzer.cargo.extraEnv": {
        "CARGO_NET_OFFLINE": "true"
    }
}
```

**Comment by Storyyeller** [comment]
> The config would look like the following in vs-code
> 
> // settings.json
> {
>     "rust-analyzer.cargo.extraEnv": {
>         "CARGO_NET_OFFLINE": "true"
>     }
> }

Thanks, it looks like the problem was that I was using "rust-analyzer.runnables.extraEnv" instead of "rust-analyzer.cargo.extraEnv".

Now RA appears to be using --offline and at least isn't endlessly holding the package lock file, although it also seems to fail complaining about missing package metadata and not offer any support. I take it vendoring sources is not sufficient to support RA in offline mode? Is this due to the issue Zicklag mentioned?

**Comment by ShoyuVanilla** [comment]
No, I guess that might be due to *dependencies of `core` and `std`*.
They are pre-built and bundled with toolchains, and linked while compilation in normal workflows like `cargo build`. So their dependencies are likely to be absent in your vendoring directory. But rust-analyzer cannot utilize those pre-built core and standard libraries, so it tries to fetch their dependencies when loading projects

**Comment by Krysme** [comment]
> No, I guess that might be due to _dependencies of `core` and `std`_. They are pre-built and bundled with toolchains, and linked while compilation in normal workflows like `cargo build`. So their dependencies are likely to be absent in your vendoring directory. But rust-analyzer cannot utilize those pre-built core and standard libraries, so it tries to fetch their dependencies when loading projects

Thanks for sharing your knowledge. So, basically rust distributed by `MSYS2` has such problem that `MSYS2` doesn't ship those dependencies of `std` and `core`, so they might need to update their distribution method.

**Comment by eirnym** [comment]
@ShoyuVanilla workaround with adding variable to `rust-analyzer.cargo.extraEnv` section under workspace settings doesn't work.

**Comment by ChayimFriedman2** [comment]
@eirnym It should work. What editor do you use? What's in the logs?

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/pull/156196 will fix this for the standard library `cargo metadata` call by bundling vendored dependencies for the standard library in the rust-src component.

**Comment by eirnym** [comment]
Sorry for late response, I see messages (for each package and version used) , while `cargo build --offline` (and check) builds correctly.

```
ERROR Received compiler message for unknown package: registry+https://github.com/rust-lang/crates.io-index#icu_normalizer_data@2.1.1
```

Sometimes registry is local registry located under `~/.cargo` folder, but fragment suffix is the same

also, I've enabled `rust-analyzer.cargo.noDeps` option, which (as written in description), won't download package metadata from repository

**Comment by ChayimFriedman2** [comment]
@eirnym First, this messages are not fatal, you can continue working normally.

Can you enable info-level logging (`RA_LOG=info`), locate the `cargo metadata` commands r-a runs (it should be in the logs), and post it here?

**Comment by thelenlucas** [comment]
@bjorn3 you are a saint, and this is going to improve my life immeasurably when 1.97 drops.
